ambry-cloud/src/main/java/com/github/ambry/cloud/RecoveryNetworkClient.java

```diff
@@ -124,134 +149,125 @@ public List<ResponseInfo> sendAndPoll(List<RequestInfo> requestsToSend, Set<Inte
-                  : page.getContinuationToken(), currRecoveryToken.getRequestUnits() + page.getRequestCharge(),
-              currRecoveryToken.getNumItems() + (currRecoveryToken.isEndOfPartitionReached() ? 0 : numItems),
-              currRecoveryToken.getNumBlobBytes() + (currRecoveryToken.isEndOfPartitionReached() ? 0
-                  : totalBlobBytesRead), nextCosmosContinuationToken == null, currRecoveryToken.getTokenCreateTime(),
-              backupStartTime, backupEndTime, lastQueryTime, firstBlobId, lastBlobId);
-          ++numPages;
-          long resultFetchtime = System.currentTimeMillis() - startTime;
-          logger.trace(
-              "Received cosmos query results page = {}, time = {} ms, RU = {}/s, numRows = {}, tokenLen = {}, isTokenNull = {}, isTokenSameAsPrevious = {}",
-              queryName, numPages, resultFetchtime, requestCharge, page != null ? page.getResults().size() : "null",
-              nextCosmosContinuationToken != null ? nextCosmosContinuationToken.length() : "null",
-              nextCosmosContinuationToken != null ? nextCosmosContinuationToken.isEmpty() : "null",
-              nextCosmosContinuationToken != null ? nextCosmosContinuationToken.equals(
-                  currRecoveryToken.getCosmosContinuationToken()) : "null");
+      // List N blobs with metadata from Azure storage from prev token position
+      ListBlobsOptions listBlobsOptions = new ListBlobsOptions()
+          .setDetails(new BlobListDetails().setRetrieveMetadata(true))
+          .setMaxResultsPerPage(azureCloudConfig.azureBlobStorageMaxResultsPerPage);
+      PagedResponse<BlobItem> response;
+      try {
+        response = azureSyncClient.createOrGetBlobStore(containerName)
+            .listBlobs(listBlobsOptions, null)
+            .iterableByPage(prevToken.getToken())
+            .iterator()
+            .next();
+        recoveryMetrics.listBlobsSuccessRate.mark();
+        clientCallback.onListBlobs(rinfo);
+      } catch (Throwable t) {
+        responseList.add(
+            new ReplicaMetadataResponseInfo(rinfo.getPartitionId(), rinfo.getReplicaType(),
+                ServerErrorCode.IO_Error,
+                ReplicaMetadataResponse.getCompatibleResponseVersion(request.getVersionId())));
+        logger.error("Failed to list blobs for Ambry partition {} in Azure Storage container {} due to {}",
+            rinfo.getPartitionId().getId(), containerName, t);
+        recoveryMetrics.listBlobsError.inc();
+        continue;
+      }
 
-          break;
+      // Extract ambry metadata
+      logger.trace("For container {}, number of blobItems from Azure = {}", containerName, response.getValue().size());
+      List<MessageInfo> messageInfoList = new ArrayList<>();
+      long bytesRead = 0, blobsRead = 0;
+      for (BlobItem blobItem: response.getValue()) {
+        MessageInfo messageInfo = getMessageInfo(blobItem);
+        if (messageInfo != null) {
```

Comment:
1. What's the else case? If messageInfo is null, what does that mean? 
2. Do you need error log for it?

Reply:
null messageInfo means there was an error extracting ambry-metadata from azure-metadata for a blob. The error is logged in getMessageInfo.